Bug: invalid character in domain in case of OpenAPI Bin creation

Hi!

I have created a few mockbins via the new "Create an OpenAPI Bin" functionality and the tool always generated this URL format:
 `https://<unique_identifier>_oas.api.mockbin.io/`

When attempting to make an HTTPS request to a URL via a simple Java application, the request fails with the following exception:
`Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure`


It seems that the root cause is that the domain `<unique_identifier>_oas` contains an underscore (_), which is not a valid character for DNS hostnames according to [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035).

Underscores are not supported in public DNS hostnames and cause TLS clients to reject the request during the handshake phase.

Requests to generated mockbin URLs should use valid domain names (only LDH characters: letters, digits, hyphens).

Could you please investigate this problem?

Thanks a lot!

--- a/config/routes.oas.json
+++ b/config/routes.oas.json
@@ -114,6 +114,27 @@
         "operationId": "get_bin"
       }
     },
+    "/v1/bins/{binId}/original.yaml": {
+      "x-zuplo-path": {
+        "pathMode": "open-api"
+      },
+      "get": {
+        "summary": "Get original YAML",
+        "description": "Returns the original YAML document uploaded when the OpenAPI bin was created",
+        "x-zuplo-route": {
+          "corsPolicy": "anything-goes",
+          "handler": {
+            "export": "getOriginalYaml",
+            "module": "$import(./modules/handlers)",
+            "options": {}
+          },
+          "policies": {
+            "inbound": ["block-bad-bins", "rate-limit-inbound"]
+          }
+        },
+        "operationId": "get_original_yaml"
+      }
+    },
     "/v1/bins/{binId}/requests": {
       "x-zuplo-path": {
         "pathMode": "open-api"

--- a/modules/handlers.ts
+++ b/modules/handlers.ts
@@ -30,6 +30,12 @@ import { default as yaml } from "./third-party/yaml/index";
 
 const MAX_SIZE = 1_048_576;
 
+function getOriginalYamlStorageKey(binId: string) {
+  return binId.endsWith("-oas")
+    ? `${binId}-YAML-original.yaml`
+    : `${binId}_YAML_original.yaml`;
+}
+
 export async function createMockResponse(
   request: ZuploRequest,
   context: ZuploContext,
@@ -153,13 +159,8 @@ async function handleOpenApiMock(
 
   let originalYamlUrl: string | undefined;
   if (isYaml) {
-    // Save the original YAML file
-    const yamlBinId = `${binId}-YAML-original`;
-    await storage.uploadObject(`${yamlBinId}.yaml`, body);
-    const yamlUrl = getInvokeBinUrl(url, yamlBinId);
-    originalYamlUrl = yamlUrl.href;
-
-    // Add x-mockbin-original-url to the parsed content
+    await storage.uploadObject(getOriginalYamlStorageKey(binId), body);
+    originalYamlUrl = `${url.origin}/v1/bins/${binId}/original.yaml`;
     parsedContent["x-mockbin-original-url"] = originalYamlUrl;
   }
 
@@ -233,6 +234,36 @@ export async function getMockResponse(
   }
 
   return responseData;
+}
+
+export async function getOriginalYaml(
+  request: ZuploRequest,
+  context: ZuploContext,
+) {
+  const { binId } = request.params;
+
+  if (!validateBinId(binId) || !isOasBin(binId)) {
+    return HttpProblems.badRequest(request, context, {
+      detail: "Invalid binId",
+    });
+  }
+
+  const storage = storageClient(context.log);
+  let response: GetObjectResult;
+  try {
+    response = await storage.getObject(getOriginalYamlStorageKey(binId));
+  } catch (err) {
+    context.log.error(err);
+    return getProblemFromStorageError(err, request, context);
+  }
+
+  return new Response(response.body, {
+    status: 200,
+    headers: {
+      "Content-Type": "application/yaml",
+      "Cache-Control": "public, max-age=31536000, immutable",
+    },
+  });
 }
 
 export async function listRequests(

--- a/modules/handlers.ts
+++ b/modules/handlers.ts
@@ -55,7 +55,7 @@ export async function createMockResponse(
     } else if (contentType.startsWith("multipart/form-data")) {
       isOpenApi = true;
       // Handle OpenAPI mock
-      binId += "_oas";
+      binId += "-oas";
       responseData = await handleOpenApiMock(
         request,
         context,
@@ -154,7 +154,7 @@ async function handleOpenApiMock(
   let originalYamlUrl: string | undefined;
   if (isYaml) {
     // Save the original YAML file
-    const yamlBinId = `${binId}_YAML_original`;
+    const yamlBinId = `${binId}-YAML-original`;
     await storage.uploadObject(`${yamlBinId}.yaml`, body);
     const yamlUrl = getInvokeBinUrl(url, yamlBinId);
     originalYamlUrl = yamlUrl.href;

--- a/modules/utils.ts
+++ b/modules/utils.ts
@@ -83,5 +83,5 @@ export function validateOpenApiDocument(document: any): void {
 }
 
 export function isOasBin(binId: string) {
-  return binId.indexOf("-oas") > 0 || binId.indexOf("_oas") > 0;
+  return binId.endsWith("-oas") || binId.endsWith("_oas");
 }

--- a/modules/utils.ts
+++ b/modules/utils.ts
@@ -45,7 +45,7 @@ export function getInvokeBinUrl(url: URL, binId: string) {
   return mockUrl;
 }
 
-const binRegEx = /^[0-9a-fA-F]{32}(_oas)?$/;
+const binRegEx = /^[0-9a-fA-F]{32}([-_]oas)?$/;
 
 export function validateBinId(binId: string) {
   return binRegEx.test(binId);
@@ -83,5 +83,5 @@ export function validateOpenApiDocument(document: any): void {
 }
 
 export function isOasBin(binId: string) {
-  return binId.indexOf("_oas") > 0;
+  return binId.indexOf("-oas") > 0 || binId.indexOf("_oas") > 0;
 }

--- a/package.json
+++ b/package.json
@@ -18,7 +18,7 @@
     "@xmldom/xmldom": "0.9.9",
     "vitest": "^2.1.4",
     "yaml": "^2.6.0",
-    "zuplo": "^6.61.23"
+    "zuplo": "6.61.23"
   },
   "devDependencies": {
     "@zuplo/checkly": "^1.2.0",

--- a/tests/bin-id.test.ts
+++ b/tests/bin-id.test.ts
@@ -55,6 +55,14 @@ describe("isOasBin", () => {
   test("returns false for plain bin", () => {
     assert.isFalse(isOasBin("12345678901234567890123456789012"));
   });
+
+  test("returns false when -oas appears mid-string but is not the suffix", () => {
+    assert.isFalse(isOasBin("foo-oas-bar"));
+  });
+
+  test("returns false when _oas appears mid-string but is not the suffix", () => {
+    assert.isFalse(isOasBin("foo_oas_bar"));
+  });
 });
 
 describe("DNS hostname compliance", () => {

--- a/tests/bin-id.test.ts
+++ b/tests/bin-id.test.ts
@@ -1,0 +1,66 @@
+import { describe, test, assert, vi, beforeAll } from "vitest";
+
+vi.mock("@zuplo/runtime", () => ({
+  HttpProblems: {},
+  ZuploContext: class {},
+  ZuploRequest: class {},
+}));
+vi.mock("../modules/env", () => ({ USE_WILDCARD_SUBDOMAIN: false }));
+vi.mock("../modules/storage", () => ({ StorageError: class {} }));
+
+let validateBinId: (id: string) => boolean;
+let isOasBin: (id: string) => boolean;
+
+beforeAll(async () => {
+  const utils = await import("../modules/utils");
+  validateBinId = utils.validateBinId;
+  isOasBin = utils.isOasBin;
+});
+
+describe("validateBinId", () => {
+  test("accepts plain 32-hex bin id", () => {
+    assert.isTrue(validateBinId("12345678901234567890123456789012"));
+  });
+
+  test("accepts new hyphen OAS suffix", () => {
+    assert.isTrue(validateBinId("12345678901234567890123456789012-oas"));
+  });
+
+  test("accepts legacy underscore OAS suffix", () => {
+    assert.isTrue(validateBinId("12345678901234567890123456789012_oas"));
+  });
+
+  test("rejects too-short id", () => {
+    assert.isFalse(validateBinId("abc"));
+  });
+
+  test("rejects non-hex chars in id", () => {
+    assert.isFalse(validateBinId("zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"));
+  });
+
+  test("rejects unknown suffix", () => {
+    assert.isFalse(validateBinId("12345678901234567890123456789012-foo"));
+  });
+});
+
+describe("isOasBin", () => {
+  test("detects new hyphen format", () => {
+    assert.isTrue(isOasBin("12345678901234567890123456789012-oas"));
+  });
+
+  test("detects legacy underscore format", () => {
+    assert.isTrue(isOasBin("12345678901234567890123456789012_oas"));
+  });
+
+  test("returns false for plain bin", () => {
+    assert.isFalse(isOasBin("12345678901234567890123456789012"));
+  });
+});
+
+describe("DNS hostname compliance", () => {
+  test("new OAS bin id contains no underscores", () => {
+    const binId = "12345678901234567890123456789012-oas";
+    assert.isFalse(binId.includes("_"));
+    assert.match(binId, /^[a-z0-9-]+$/);
+  });
+});

--- a/www/pages/bins/[binId].tsx
+++ b/www/pages/bins/[binId].tsx
@@ -235,7 +235,8 @@ const Bin = () => {
 
   const docsUrl = `https://zudoku.dev/demo?api-url=${process.env.NEXT_PUBLIC_API_URL}/v1/bins/${binId}`;
   const binUrl = requests.url ?? `${process.env.NEXT_PUBLIC_API_URL}/${binId}`;
-  const isOas = (binId ?? "").indexOf("_oas") > 0;
+  const isOas =
+    (binId ?? "").indexOf("-oas") > 0 || (binId ?? "").indexOf("_oas") > 0;
 
   return (
     <main className="bg-bg-subtle min-h-screen">

--- a/www/pages/bins/[binId].tsx
+++ b/www/pages/bins/[binId].tsx
@@ -235,8 +235,8 @@ const Bin = () => {
 
   const docsUrl = `https://zudoku.dev/demo?api-url=${process.env.NEXT_PUBLIC_API_URL}/v1/bins/${binId}`;
   const binUrl = requests.url ?? `${process.env.NEXT_PUBLIC_API_URL}/${binId}`;
-  const isOas =
-    (binId ?? "").endsWith("-oas") || (binId ?? "").endsWith("_oas");
+  const binIdStr = typeof binId === "string" ? binId : "";
+  const isOas = binIdStr.endsWith("-oas") || binIdStr.endsWith("_oas");
 
   return (
     <main className="bg-bg-subtle min-h-screen">

--- a/www/pages/bins/[binId].tsx
+++ b/www/pages/bins/[binId].tsx
@@ -236,7 +236,7 @@ const Bin = () => {
   const docsUrl = `https://zudoku.dev/demo?api-url=${process.env.NEXT_PUBLIC_API_URL}/v1/bins/${binId}`;
   const binUrl = requests.url ?? `${process.env.NEXT_PUBLIC_API_URL}/${binId}`;
   const isOas =
-    (binId ?? "").indexOf("-oas") > 0 || (binId ?? "").indexOf("_oas") > 0;
+    (binId ?? "").endsWith("-oas") || (binId ?? "").endsWith("_oas");
 
   return (
     <main className="bg-bg-subtle min-h-screen">

